File: plot_arc.m
function plot_arc(x, y, theta, label, arrow)

if nargin<5
    arrow=1;
end

EPS = .000001;
NUM_LINES = 30;
ARROW_T = pi/12;
ARROW_R = .05;
ARROW_OFF = .6;
LABEL_OFF = .01;

vf = [diff(x); diff(y)];
if vf(2)/vf(1)<0
    theta = -theta;
end
if vf(2)==0
    theta = 0;
end

S = norm(vf);
k_inv = S/(2*sin(theta)+EPS);

s = k_inv*tan(2*theta/NUM_LINES);

fix = 1;
if S==0
    fix = 0;
    s = .5/NUM_LINES;
    theta = pi;
end

if theta==0
    s = S/NUM_LINES;
end

cn = cos(-2*theta/(NUM_LINES-1));
sn = sin(-2*theta/(NUM_LINES-1));
Rn = [cn, -sn; sn, cn];     

c0 = cos(theta);
s0 = sin(theta);
R0 = [c0, -s0; s0, c0];
v0 = R0*(s*[diff(x); diff(y)]/S);

ai = round(NUM_LINES*ARROW_OFF);
X = zeros(1, NUM_LINES+1);
Y = zeros(1, NUM_LINES+1);
X(1) = x(1); Y(1) = y(1);
v = v0;
for i=1:NUM_LINES
    X(i+1) = X(i) + v(1);
    Y(i+1) = Y(i) + v(2);
    v = Rn*v;
    if i==ai
        vf = ARROW_R*v/norm(v); % arrow direction
    end
end
if fix
    Sf = sqrt((X(end)-X(1))^2 + (Y(end)-Y(1))^2);
    X = X(1) + S*(X-X(1))/Sf;
    Y = Y(1) + S*(Y-Y(1))/Sf;
end

hold on
plot(X,Y,'r-');
ca = cos(ARROW_T);
sa = sin(ARROW_T);
Ra = [ca, -sa; sa, ca];

ai = ai+2;
label_off = round(NUM_LINES*LABEL_OFF/s);
if nargin>=4        
    xt = X(ai-label_off);
    yt = Y(ai-label_off);
    text(xt, yt, label, 'BackgroundColor',[.9 .9 .9], 'VerticalAlignment', 'Middle', 'HorizontalAlignment', 'Center')
end

if arrow
f1 = [X(ai); Y(ai)]-Ra*vf;
f2 = [X(ai); Y(ai)]-Ra\vf;
fill([X(ai), f1(1), f2(1)], [Y(ai), f1(2), f2(2)], 'r')
end
hold off

end

File: python/plot_arc.py

import numpy as np
import scipy
import matcompat

# if available import pylab (from matlibplot)
try:
    import matplotlib.pylab as plt
except ImportError:
    pass

def plot_arc(x, y, theta, label, arrow):

    # Local Variables: f1, f2, cn, vf, ai, ca, EPS, Ra, Rn, xt, LABEL_OFF, fix, s0, label, theta, R0, v0, S, k_inv, Y, X, c0, ARROW_OFF, Sf, ARROW_T, NUM_LINES, ARROW_R, i, label_off, s, sn, arrow, v, y, x, sa, yt
    # Function calls: plot, cos, plot_arc, text, sqrt, round, nargin, zeros, diff, tan, pi, sin, norm, fill
    if nargin<5.:
        arrow = 1.
    
    
    EPS = .000001
    NUM_LINES = 30.
    ARROW_T = np.pi/12.
    ARROW_R = .05
    ARROW_OFF = .6
    LABEL_OFF = .01
    vf = np.array(np.vstack((np.hstack((np.diff(x))), np.hstack((np.diff(y))))))
    if matdiv(vf[1], vf[0])<0.:
        theta = -theta
    
    
    if vf[1] == 0.:
        theta = 0.
    
    
    S = linalg.norm(vf)
    k_inv = matdiv(S, 2.*np.sin(theta)+EPS)
    s = np.dot(k_inv, np.tan(matdiv(2.*theta, NUM_LINES)))
    fix = 1.
    if S == 0.:
        fix = 0.
        s = matdiv(.5, NUM_LINES)
        theta = np.pi
    
    
    if theta == 0.:
        s = matdiv(S, NUM_LINES)
    
    
    cn = np.cos(matdiv(np.dot(-2., theta), NUM_LINES-1.))
    sn = np.sin(matdiv(np.dot(-2., theta), NUM_LINES-1.))
    Rn = np.array(np.vstack((np.hstack((cn, -sn)), np.hstack((sn, cn)))))
    c0 = np.cos(theta)
    s0 = np.sin(theta)
    R0 = np.array(np.vstack((np.hstack((c0, -s0)), np.hstack((s0, c0)))))
    v0 = np.dot(R0, matdiv(np.dot(s, np.array(np.vstack((np.hstack((np.diff(x))), np.hstack((np.diff(y))))))), S))
    ai = np.round(np.dot(NUM_LINES, ARROW_OFF))
    X = np.zeros(1., (NUM_LINES+1.))
    Y = np.zeros(1., (NUM_LINES+1.))
    X[0] = x[0]
    Y[0] = y[0]
    v = v0
    for i in np.arange(1., (NUM_LINES)+1):
        X[int((i+1.))-1] = X[int(i)-1]+v[0]
        Y[int((i+1.))-1] = Y[int(i)-1]+v[1]
        v = np.dot(Rn, v)
        if i == ai:
            vf = matdiv(np.dot(ARROW_R, v), linalg.norm(v))
            #% arrow direction
        
        
        
    if fix:
        Sf = np.sqrt(((X[int(0)-1]-X[0])**2.+(Y[int(0)-1]-Y[0])**2.))
        X = X[0]+matdiv(np.dot(S, X-X[0]), Sf)
        Y = Y[0]+matdiv(np.dot(S, Y-Y[0]), Sf)
    
    
    plt.hold(on)
    plt.plot(X, Y, 'r-')
    ca = np.cos(ARROW_T)
    sa = np.sin(ARROW_T)
    Ra = np.array(np.vstack((np.hstack((ca, -sa)), np.hstack((sa, ca)))))
    ai = ai+2.
    label_off = np.round(matdiv(np.dot(NUM_LINES, LABEL_OFF), s))
    if nargin >= 4.:
        xt = X[int((ai-label_off))-1]
        yt = Y[int((ai-label_off))-1]
        plt.text(xt, yt, label, 'BackgroundColor', np.array(np.hstack((.9, .9, .9))), 'VerticalAlignment', 'Middle', 'HorizontalAlignment', 'Center')
    
    
    if arrow:
        f1 = np.array(np.vstack((np.hstack((X[int(ai)-1])), np.hstack((Y[int(ai)-1])))))-np.dot(Ra, vf)
        f2 = np.array(np.vstack((np.hstack((X[int(ai)-1])), np.hstack((Y[int(ai)-1])))))-linalg.solve(Ra, vf)
        plt.fill(np.array(np.hstack((X[int(ai)-1], f1[0], f2[0]))), np.array(np.hstack((Y[int(ai)-1], f1[1], f2[1]))), 'r')
    
    
    plt.hold(off)
    return 